### ISSUE TITLE
FileMerger: add option for non-overlapping RTs

### ISSUE THREAD

**Issue by hendrikweisser** [OpenMS/OpenMS#1500]
FileMerger currently merges files by simply accumulating all relevant elements (spectra, features) in one container. This makes it difficult to process the merged data in a way that allows results to be related back to the original "chunks".

For example, if biological samples are fractionated prior to MS analysis, it can be desirable to merge the mzML files of different fractions before running database searches, to save time and perhaps improve statistical analyses (e.g. Percolator) by increasing sample sizes. However, e.g. for quantification it would also be desirable to relate the search results back to the original fractions. With the current merge strategy this doesn't seem possible.

A simple solution would be to adjust the retention time data (RTs) of elements, so that different chunks (from different input files) have non-overlapping RT ranges. After merging, the chunks would appear consecutively in RT. There should also be an option to insert a time gap of a user-definable length between the chunks (to avoid artifacts at the boundaries).

**Comment by timosachsenberg** [comment]
I am by any means not a fan of this approach as it missuses the standard formats and destroys a lot of tracked meta information.
That beeing said: MapRTTransformer seems for me to best place to put the transformation before merging.

**Comment by hendrikweisser** [comment]
> I am by any means not a fan of this approach as it missuses the standard formats and destroys a lot of tracked meta information.

Losing meta information is of course a valid concern, but compared to the way we currently merge, all that is lost is the original RT information (and it would be possible to store the RT offsets of the chunks as meta data, allowing to reconstruct the original RTs).

Would you have another suggestion for how to solve the problem in the example? Unfortunately most types of meta data in an mzML would be lost either during the search (protein identification) or when search results are imported back into OpenMS.

**Comment by hendrikweisser** [comment]
> That beeing said: MapRTTransformer seems for me to best place to put the transformation before merging.

Since the transformation depends on the other maps being merged, doing it via MapRTTransformer for each map would involve too much manual work (finding the RT ranges of maps and generating trafoXMLs). I think it would be a relatively simple addition to FileMerger.

**Comment by hroest** [comment]
I am not on principle against Hendriks suggestion as it is an option only.
However I wonder how many people will then actually use the proposed
workflow as it seems nontrivial and abuse the mzML format a bit.

@Hendrik: wouldnt the "proper" way to deal with this situation be to *not*
merge the mzMLs but keep them as separate files on your disk, one file for
each fractionation? This way you could have the original retention times
and the information which file was which. If you want to improve
statistical analysis, I think it should be the job of the percolator and
TPP to allow multiple input files which I think at least TPP already does.

Anyways, I dont feel it is a very elegant solution and Hendrik could think
about whether this is *really* really what he wants. Just to put this out
there, in the mzML standard it specifically says that there is only 1 <run>
tag and "A run in mzML should correspond to a single, consecutive and
coherent set of scans on an instrument."
http://www.peptideatlas.org/tmp/mzML1.1.0.html#sampleList -- so maybe
putting all spectra into one single mzML is really not what mzML intended
to do. It was specifically intended to represent a single MS run and you
really *should* use multiple mzML files for multiple runs if at all
possible.

Regarding the main arguments, I dont think (i) saving time is a good
argument as it is not clear to me whether you really save time with one big
file and regarding (ii) the statistical analysis, this seems to be a
shortcoming of the tools if they cannot handle multiple mzML files (and
some like the TPP do).

Best

Hannes

**Comment by hendrikweisser** [comment]
@timosachsenberg, @hroest:
I think this discussion is really useful, so thanks.

I'm seeing the merging of mzMLs from fractions as just an intermediate step, to simplify searching and allow/improve re-scoring of search results (in our case with Percolator, which AFAIK needs at least ~1000 hits to work reliably). If we make RT adjustments during merging, it would certainly be possible to split up the search results and undo the RT transformations, in order to then have several idXMLs that correspond to the different fractions (original mzMLs).
However, if we are anyway interested in results for all fractions together, this may not even be necessary/useful.

Regarding the abuse of mzML: I admit that merging different runs into one file apparently violates the specification of the format. However, I don't see it as a big deal since the merged mzML is just an intermediate artifact and the original files are not lost. In addition, analyzing multiple fractions together is simply not supported by many tools (e.g. OpenMS, Mascot, Percolator...). However, this usually isn't a big problem since input or output files of multiple fractions can be merged instead. (And since this is easy, I don't think it would be worth to add the overhead of dealing with multiple files from different fractions in the respective tools.)

**Comment by timosachsenberg** [comment]
I think it is fine with me if it is documented and an advanced option. So how would you suggest would the retransformation work in practice?

**Comment by hendrikweisser** [comment]
> I think it is fine with me if it is documented and an advanced option.

I can make it advanced if you really want, but since the problematic aspects are inherent to FileMerger itself, I don't understand why.

> So how would you suggest would the retransformation work in practice?

Good question. FileMerger would need to write out the necessary information - which RT regions in the merged file correspond to which original RT ranges. Then we'd need to make a UTILS tool that takes this information along with a merged input file (potentially derived from an earlier merged file, e.g. an idXML resulting from a search of a merged mzML) and splits the input file.
I see two alternatives for storing the RT information: In the merged file as meta values, which we should probably do in any case - but this has the disadvantage that we might have to read in a huge mzML in order to split e.g. an idXML. Or alternatively (in addition?) in trafoXML, either in one file or in one per original input (e.g. fraction).

Thoughts?

**Comment by hroest** [comment]
@hendrikweisser if we go ahead with this, I would prefer the solution to store two meta values in each spectrum specifying (i) to which fraction / mzML filename it belonged and (ii) the original RT. Then you can change the RT field the way you want. But I would prefer if the "real" information needed for the back-transformation is stored explicitly. In this manner, every tool reading the large mzML will not be impacted but it would be relatively straightforward for OpenMS to retrieve the "real" information and piece together the original files. I think the solution with extra files makes everything quite a bit harder and is at least (if not more) unportable than my suggestion. I would prefer to keep the information together and to keep it in a way that a human reading the XML file can understand what is going on.

I dont think reading the huge mzML file is a problem any more, especially with all the improvements that I have made in reading such large files. You can either stream them now and from the stream only keep the ones you are interested in or you can use an indexed version of the file (and keep the "empty" spectra without the data in memory around so you know which indexed spectra you are interested in). See also http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0125108

**Comment by hendrikweisser** [comment]
> if we go ahead with this, I would prefer the solution to store two meta values in each spectrum specifying (i) to which fraction / mzML filename it belonged and (ii) the original RT.

I can do this, no problem, I just want to point out that currently:
- we don't store the original filename for spectra that are merged by FileMerger
- we don't store the original RT when RTs are adjusted by MapRTTransformer or a MapAligner

> I think the solution with extra files makes everything quite a bit harder and is at least (if not more) unportable than my suggestion.

Just storing the original filenames and RTs of spectra doesn't directly give us the information that would be needed to split up e.g. an idXML file from a database search of a merged mzML (at least not in a convenient way). However, it should be easy to generate a trafoXML file for every fraction (mzML file) that maps it to its RT region in the merged file - such that the following pipelines would be equivalent (produce the same merged file):
- `FileMerger -in a.mzML b.mzML -raw:rt_concat 1 -out merged.mzML -trafo_out a.trafoXML b.trafoXML` ("raw:rt_concat 1" concatenates the files with a 1 second RT gap)
- `MapRTTransformer -in a.mzmL -trafo_in a.trafoXML -out a2.mzML; MapRTTransformer -in b.mzML -trafo_in b.trafoXML -out b2.mzML; FileMerger -in a2.mzML b2.mzML -out merged.mzML`

To get the original files back, we could envision something like:
`FileUnmerger -in merged.mzML -trafo_in a.trafoXML b.trafoXML -out a.mzML b.mzML` (although in this case passing the trafoXMLs would be optional, since the necessary information is stored in meta values). This approach would give us flexibility to extract only some of the original fractions from the merged file.

**Comment by hroest** [comment]
I think I can go with the proposed workflow, but I dont quite see how 
```
FileUnmerger -in merged.mzML -trafo_in a.trafoXML b.trafoXML -out a.mzML b.mzML
```
would really know which fractions to extract, especially if you only wanted b and not a (would you somehow encode the name of the file or how would you address individual fractions?). Personally, I think the feature "only extract certain fractions" is nice to have but would not be first priority. Simply having `FileMerger` and `FileSplitter` (maybe a better name than FileUnmerger) functionality that undoes the merger operation would be a good first step. So in total I like

```
FileMerger -in a.mzML b.mzML -raw:rt_concat 1 -out merged.mzML -trafo_out a.trafoXML b.trafoXML
FileUnmerger -in merged.mzML -trafo_in a.trafoXML b.trafoXML -out a.mzML b.mzML 
```

although I am not sure whether messing with all the trafo files will not make things very complicated and I think that also 

```
FileUnmerger -in merged.mzML -out a.mzML b.mzML 
```

should be something that works. Especially as I often look at raw XML files (I dont know whether you do as well :-) I think writing the information in a more or less human-understandable format helps a lot in avoiding bugs and making sure people who work with the files understand what type of files they are. I think in this respect (human readability, not all others), the TPP file formats like pepxml are quite good as you can easily see in a large pepxml where each spectrum came from originally. Another idea would be to simply encode it in the spectrum native id but I am against this approach as it will not be machine readable.

**Comment by hendrikweisser** [comment]
> I think I can go with the proposed workflow, but I dont quite see how [...] would really know which fractions to extract, especially if you only wanted b and not a (would you somehow encode the name of the file or how would you address individual fractions?).

Based on the trafoXML! Each trafoXML stores two RT pairs: (old_rt_start, new_rt_start), (old_rt_end, new_rt_end). To get the original fraction back, we simply extract everything between "new_rt_start" and "new_rt_end" (and then adjust the RTs).

>  I think that also `FileUnmerger -in merged.mzML -out a.mzML b.mzML` should be something that works.

That could be done for the mzML, but remember that the original use case was to split up an idXML that was derived from a merged mzML. There we wouldn't have the meta values in the file - they don't get carried over -, so we'd have to rely on secondary input files to tell us how to split.
Also, a problem with the command that you wrote is that the correct number of output filenames needs to be supplied to "out", and it may not be immediately clear what that number is. (There are workarounds for this, of course, like giving an output directory instead of filenames.)

**Comment by hroest** [comment]
true true

> Based on the trafoXML! Each trafoXML stores two RT pairs: (old_rt_start, new_rt_start), (old_rt_end, new_rt_end). To get the original fraction back, we simply extract everything between "new_rt_start" and "new_rt_end" (and then adjust the RTs).

true, I did not consider that. 

> That could be done for the mzML, but remember that the original use case was to split up an idXML that was derived from a merged mzML. There we wouldn't have the meta values in the file - they don't get carried over -, so we'd have to rely on secondary input files to tell us how to split.

is this a limitation of idXML or of OpenMS? I am not 100% clear on how idXML works but AFAIK there should be some unique identifier to which spectrum an id refers to - so here we should ensure that there are stable ids in the original mzML and the merged mzML if at all possible

well anyways, I think we are somewhat clear on how the solution should look like and I propose the next step is implementation ...

**Comment by timosachsenberg** [comment]
I will work on saving the original MS run location in featuremaps/XML as well as in identifications + idXML and check if consensusmap/XML supports this.
With the native spectrum id tracing back the spectrum should be possible.

**Comment by hendrikweisser** [comment]
> I am not 100% clear on how idXML works but AFAIK there should be some unique identifier to which spectrum an id refers to - so here we should ensure that there are stable ids in the original mzML and the merged mzML if at all possible

There is no such field in idXML. There is RT and precursor m/z information, though, which can be used to find the originating spectrum.

**Comment by timosachsenberg** [comment]
There is an spectrum_reference identifier
```
<xs:attribute name="spectrum_reference" type="xs:unsignedInt">
  <xs:annotation>
  <xs:documentation>Integer reference number of the identified spectrum (or feature </xs:documentation>
  </xs:annotation>
</xs:attribute>
```

**Comment by timosachsenberg** [comment]
It's not the native id (which would be a string) but maybe it is sufficient

**Comment by hendrikweisser** [comment]
> There is an spectrum_reference identifier

Interesting. I don't think I've seen this in any idXML file "in the wild" yet. We should check which of our search engine adapters actually support this.

**Comment by hroest** [comment]
having some way of referencing where an id came from seems like a good idea anyways. 

> I will work on saving the original MS run location in featuremaps/XML as well as in identifications + idXML and check if consensusmap/XML supports this.
> With the native spectrum id tracing back the spectrum should be possible.

@timosachsenberg that sounds like a good idea to store the original filename as well somewhere - it will still be somewhat hairy as "original file id + original RT + original m/z" is still only unambiguous if the RT is transformed in the "correct" way according to Hendriks scheme but at least it seems workable. If we use the `spectrum_reference` as an int, it will probably be the spectrum number  from the merged file, correct?

**Comment by timosachsenberg** [comment]
for mzML the spectrum id is just the original spectrum number (before anything got removed in filtering steps etc.). So it is crucial be to retain this original spectrum id and link to the ms file through any processing.

**Comment by hendrikweisser** [comment]
Apropos native IDs: FileMerger with mzML input currently rewrites the native IDs and just assigns consecutive numbers. This leads to a warning when storing the output mzML:
> Invalid native IDs detected. Using spectrum identifier nativeID format (spectrum=xsd:nonNegativeInteger) for all spectra.

Then, the native IDs are rewritten again in the format "spectrum=N" by the MzMLHandler.

This behaviour obviously isn't very clever. The question is, what would be the best alternative:
1. Keep the original native IDs, and risk having duplicates in the merged file.
2. Amend the original native IDs to make them unique (e.g. add "FileMerger_index=N" for the N'th input file).
3. Keep the consecutive numbering scheme, but use the format "spectrum=N" already in FileMerger.

I could also add a parameter to let the user decide, but I'd prefer if we could agree on one method.

**Comment by timosachsenberg** [comment]
I think keeping the native ID is as intended by thr standard. Native ID + source File should be a unique combination in most cases

**Comment by hendrikweisser** [comment]
> Native ID + source File should be a unique combination in most cases

Okay. Then we need to check that the source file annotations are handled correctly in FileMerger, which I'm not sure about. The code is in line 391 and following of the new FileMerger; I haven't changed the logic of it. In particular, it seems that individual spectra are only annotated with source file information if there's only one spectrum in an input - which may make sense for single-spectrum DTA input files, but perhaps not for mzML.

**Comment by hendrikweisser** [comment]
I've opened a new ticket for the "FileUnmerger"/"FileSplitter" utility (#1541). Closing this one as the main issue is resolved via #1539.
